projects/Unit/TestAsyncConsumerExceptions.cs

```diff
@@ -40,20 +40,89 @@ namespace RabbitMQ.Client.Unit
     [TestFixture]
     public class TestAsyncConsumerExceptions : IntegrationFixture
     {
+        protected void TestExceptionHandlingWith(IBasicConsumer consumer,
+            Action<IModel, string, IBasicConsumer, string> action)
+        {
+            object o = new object();
+            bool notified = false;
+            string q = _model.QueueDeclare();
+
+
+            _model.CallbackException += (m, evt) =>
+            {
+                notified = true;
+                Monitor.PulseAll(o);
+            };
+
+            string tag = _model.BasicConsume(q, true, consumer);
+            action(_model, q, consumer, tag);
+            WaitOn(o);
+
+            Assert.IsTrue(notified);
+        }
+
+        [SetUp]
+        public override void Init()
+        {
+            _connFactory = new ConnectionFactory
+            {
+                DispatchConsumersAsync = true
+            };
+
+            _conn = _connFactory.CreateConnection();
+            _model = _conn.CreateModel();
+        }
+
+        [Test]
+        public void TestCancelNotificationExceptionHandling()
+        {
+            IBasicConsumer consumer = new ConsumerFailingOnCancel(_model);
+            TestExceptionHandlingWith(consumer, (m, q, c, ct) => m.QueueDelete(q));
+        }
+
+        [Test]
+        public void TestConsumerCancelOkExceptionHandling()
+        {
+            IBasicConsumer consumer = new ConsumerFailingOnCancelOk(_model);
+            TestExceptionHandlingWith(consumer, (m, q, c, ct) => m.BasicCancel(ct));
+        }
+
+        [Test]
+        public void TestConsumerConsumeOkExceptionHandling()
+        {
+            IBasicConsumer consumer = new ConsumerFailingOnConsumeOk(_model);
+            TestExceptionHandlingWith(consumer, (m, q, c, ct) => { });
+        }
+
+        [Test]
+        public void TestConsumerShutdownExceptionHandling()
+        {
+            IBasicConsumer consumer = new ConsumerFailingOnShutdown(_model);
+            TestExceptionHandlingWith(consumer, (m, q, c, ct) => m.Close());
+        }
+
+        [Test]
+        public void TestDeliveryExceptionHandling()
+        {
+            IBasicConsumer consumer = new ConsumerFailingOnDelivery(_model);
+            TestExceptionHandlingWith(consumer, (m, q, c, ct) => m.BasicPublish("", q, null, _encoding.GetBytes("msg")));
+        }
+
         private class ConsumerFailingOnDelivery : AsyncEventingBasicConsumer
         {
             public ConsumerFailingOnDelivery(IModel model) : base(model)
             {
             }
 
-            public override Task HandleBasicDeliver(string consumerTag,
+            public override async Task HandleBasicDeliver(string consumerTag,
                 ulong deliveryTag,
                 bool redelivered,
                 string exchange,
                 string routingKey,
                 IBasicProperties properties,
                 ReadOnlyMemory<byte> body)
             {
+                await Task.Delay(0);
```

Comment:
I find this hard to understand because the thing you intend to test is the throwing on the synchronous path, but then there is still the async statemachine at play and the reader if the test need to now know that Task Delay with zero always returns a completed task and therefore sync completes. Still though the presence of the async key work essentially changes the exception semantics at play

Reply:
https://dotnetfiddle.net/1DUM2W

You can observe the difference by executing the above fiddle. Once as is and once with the async keyword on Throws

Reply:
The purpose of those tests is to reproduce the bug #944. For cases which the async change the exception semantics.
This line needed only to clean the compile warning, not changing the functionality. The other option was suppress  the warning, which lead to more confusing then this.
Or calling another random async function and await, but still - it's a random action just to remove the warning.

Reply:
My point was any async / await code that yields you add changes the behaviour of the code. The description of the issue you reference talks about sync completion which is the case when IsCompleted is true right after scheduling. It is almost impossible to run with yielding into the sync completion case there you would always land on the path that already today awaits the task. Or in other words are you sure the test you wrote are correctly testing the else part?

Reply:
You could use Task.FromException instead if I'm not mistaken 
https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.fromexception?view=netcore-3.1

Reply:
Yes, it tests what it should be. The test classes throws exception without waiting, so with an average CPU it will get task done flag without a problem.
Changed to Task.FromException anyway, it's more elegant way.